### PR TITLE
[FIP-37] Add bitmap scalar functions and register via FlussCatalog

### PR DESCRIPTION
**Pull request by pbanakar**
### Purpose

Linked issue: Part of #3289

This PR completes the Phase 1 implementation of FIP-37 by adding the six RoaringBitmap scalar functions. It builds on the infrastructure from PR #3319  and the aggregate functions from PR #3398 

These functions are registered as built-in functions in `FlussCatalog`, allowing users to call them directly in Flink SQL without manual `CREATE FUNCTION` statements.

### Brief change log

**New files in `fluss-flink/fluss-flink-common`:**

*   `RbCardinalityFunction.java`: `rb_cardinality(BYTES) -> BIGINT` — Returns the number of distinct integers in the bitmap.
*   `RbBuildFunction.java`: `rb_build(ARRAY<INT>) -> BYTES` — Constructs a bitmap from an array of integers.
*   `RbContainsFunction.java`: `rb_contains(BYTES, INT) -> BOOLEAN` — Checks if a bitmap contains a specific integer.
*   `RbToArrayFunction.java`: `rb_to_array(BYTES) -> ARRAY<INT>` — Converts a bitmap into an array of its integers.
*   `RbOrFunction.java`: `rb_or(BYTES, BYTES) -> BYTES` — Computes the union of two bitmaps.
*   `RbAndFunction.java`: `rb_and(BYTES, BYTES) -> BYTES` — Computes the intersection of two bitmaps.
*   `RbScalarFunctionsTest.java`: Unit tests for all six scalar functions, covering correctness, null handling, and edge cases.

**Modified files:**

*   `FlinkCatalog.java`: Registers the six new scalar functions in the `BUILTIN_BITMAP_FUNCTIONS` map.
*   `FlinkCatalogTest.java`: Adds a unit test to verify the scalar functions are correctly registered.
*   `RbFunctionsCatalogITCase.java`: Adds a new integration test (`testScalarFunctions`) that exercises all scalar functions through a `TableEnvironment`, validating the end-to-end catalog registration and execution path.

### Tests

*   **Unit tests:** `RbScalarFunctionsTest` contains comprehensive tests for each new scalar function.
*   **Catalog integration:** `FlinkCatalogTest` verifies that the functions are discoverable through the catalog API.
*   **End-to-end SQL ITCase:** `RbFunctionsCatalogITCase` now includes a test that runs SQL queries with `rb_cardinality`, `rb_build`, `rb_contains`, `rb_to_array`, `rb_or`, and `rb_and` against a live Flink cluster.

**Verified locally:**

*   `./mvnw spotless:apply` — 0 violations
*   `./mvnw test -pl fluss-flink/fluss-flink-common -Dtest="RbScalarFunctionsTest,FlinkCatalogTest"` — BUILD SUCCESS
*   `./mvnw verify -pl fluss-flink/fluss-flink-common -Dit.test=RbFunctionsCatalogITCase` — BUILD SUCCESS

### API and Format

This change is purely additive and does not affect any storage formats or wire protocols. The functions operate on `BYTES` columns containing standard RoaringBitmap serialized data.

### Documentation

User-facing documentation will be created in a follow-up PR now that the full set of Phase 1 functions is complete.

### CODE REVIEW

**Comment by pbanakar** [comment]
Hi @wuchong @polyzos @platinumhamburg could you please help review this!?

File: fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/functions/bitmap/RbOrFunction.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+     */
+    @Nullable
+    public byte[] eval(@Nullable byte[] leftBytes, @Nullable byte[] rightBytes) throws IOException {
+        if (leftBytes == null && rightBytes == null) {
+            return null;
+        }
+        if (leftBytes == null) {
+            return rightBytes;
+        }
+        if (rightBytes == null) {
```
Comment (platinumhamburg) [inline fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/functions/bitmap/RbOrFunction.java]: rb_or applies the aggregate-style NULL-as-ignored behavior to a scalar function. FIP-37 explicitly documents that rb_or_agg ignores NULL inputs, but also distinguishes NULL from empty bitmap. Unless scalar rb_or is specified to treat NULL as identity, this should probably return NULL when either argument is NULL, or the scalar NULL semantics need to be documented and tested.

File: fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/functions/bitmap/RbBuildFunction.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * if all inputs are null or no inputs are provided.
+ */
+public class RbBuildFunction extends ScalarFunction {
+
+    /**
+     * @param values variadic integer values to add to the bitmap; null values are ignored
+     * @return serialized bitmap, or null if all inputs are null
+     */
+    @Nullable
+    public byte[] eval(@Nullable Integer... values) throws IOException {
```
Comment (platinumhamburg) [inline fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/functions/bitmap/RbBuildFunction.java]: It looks like the implementation does not match the advertised signature for `rb_build`.

The PR description says `rb_build(ARRAY<INT>) -> BYTES`, but the UDF currently exposes:
```java
public byte[] eval(@Nullable Integer... values)
```
In Flink UDF type inference, Java varargs are treated as a varying sequence of arguments rather than a single ARRAY argument. So this supports SQL like:
```sql
rb_build(1, 2, 3)
```
but it is different from:
```sql
rb_build(ARRAY[1, 2, 3])
```
If the intended API is rb_build(ARRAY<INT>) -> BYTES, could we change the implementation to accept a non-varargs array/list-style parameter and add an SQL test for rb_build(ARRAY[1, 2, 3]) or an array column? Otherwise, the PR description and function Javadoc should document the variadic signature instead.

**Review by platinumhamburg — COMMENTED** [review]
@Prajwal-banakar Thanks for the great work. The PR LGTM. I just have a few minor questions on specific points.

File: fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/functions/bitmap/RbOrFunction.java (hunk 1 above)
Comment (pbanakar) [reply fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/functions/bitmap/RbOrFunction.java]: I've updated rb_or to return NULL if either input is NULL, aligning it with standard SQL behavior for scalar functions. The corresponding unit tests have been updated to verify this.

File: fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/functions/bitmap/RbBuildFunction.java (hunk 2 above)
Comment (pbanakar) [reply fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/functions/bitmap/RbBuildFunction.java]: I've corrected the signature to align with the documented rb_build(ARRAY<INT>) behavior. The eval method now accepts Integer[], and I've added an integration test case in RbFunctionsCatalogITCase using rb_build(ARRAY[...]) to confirm it works as expected.

**Comment by pbanakar** [comment]
Hi @platinumhamburg Thanks for the feedback, adressed the comments PTAL another look when you  have some time, Thanks again!

**Comment by platinumhamburg** [comment]
> Hi @platinumhamburg Thanks for the feedback, adressed the comments PTAL another look when you have some time, Thanks again!

@Prajwal-banakar Thanks, LGTM. @polyzos @wuchong Would you like to take another look?

**Comment by polyzos** [comment]
@Prajwal-banakar I will merge once the CI passes. Do we need any documentation here?

**Comment by pbanakar** [comment]
Hi @polyzos, no docs needed to block this merge. All three Phase 1 PRs committed to a follow-up documentation PR once the full function set was complete, and that's what I'll open next.

For the documentation itself, I see two approaches and wanted your input on which fits better:

1. Add a new section to the existing website/docs/table-design/merge-engines/aggregation.md under a "RoaringBitmap SQL Functions" heading. Simpler, keeps everything bitmap-related in one place.

2. Create a new standalone page website/docs/table-design/bitmap-functions.md. Cleaner separation since the FIP-37 functions are Flink SQL UDFs registered via FlussCatalog, which is a different concept from the server-side rbm32/rbm64 merge engine aggregators already documented there.

Worth noting that @platinumhamburg 's  [UV deduplication](https://fluss.apache.org/blog/roaringbitmap-uv-deduplication/) blog post already covers the end-to-end workflow well, so the docs page can reference that and stay focused on the function reference (signatures, null semantics, examples).

Happy to go with whichever approach you prefer, just let me know!

**Comment by polyzos** [comment]
@Prajwal-banakar, what I'm thinking - if it makes sense to you - is to go with option 1 and then just create a small follow-up blog if you want to, just to show your work and explain how it differs now in terms of "how it works".

Whatever works best for you, happy to support either
